wpilibc/src/main/native/include/frc/estimator/KalmanFilter.h

```diff
@@ -0,0 +1,230 @@
+/*----------------------------------------------------------------------------*/
+/* Copyright (c) 2018-2020 FIRST. All Rights Reserved.                        */
+/* Open Source Software - may be modified and shared by FRC teams. The code   */
+/* must be accompanied by the FIRST BSD license file in the root directory of */
+/* the project.                                                               */
+/*----------------------------------------------------------------------------*/
+
+#pragma once
+
+#include <array>
+#include <cmath>
+
+#include <Eigen/Core>
+#include <drake/math/discrete_algebraic_riccati_equation.h>
+#include <units/time.h>
+
+#include "frc/StateSpaceUtil.h"
+#include "frc/system/Discretization.h"
+#include "frc/system/LinearSystem.h"
+
+namespace frc {
+
+/**
+ * Luenberger observers combine predictions from a model and measurements to
+ * give an estimate of the true system state.
+ *
+ * Luenberger observers use an L gain matrix to determine whether to trust the
+ * model or measurements more. Kalman filter theory uses statistics to compute
+ * an optimal L gain (alternatively called the Kalman gain, K) which minimizes
+ * the sum of squares error in the state estimate.
+ *
+ * Luenberger observers run the prediction and correction steps simultaneously
+ * while Kalman filters run them sequentially. To implement a discrete-time
+ * Kalman filter as a Luenberger observer, use the following mapping:
+ * <pre>C = H, L = A * K</pre>
+ * (H is the measurement matrix).
+ *
+ * For more on the underlying math, read
+ * https://file.tavsys.net/control/state-space-guide.pdf.
+ */
+template <int States, int Inputs, int Outputs>
+class KalmanFilter {
+ public:
+  /**
+   * Constructs a state-space observer with the given plant.
+   *
+   * @param plant              The plant used for the prediction step.
+   * @param stateStdDevs       Standard deviations of model states.
+   * @param measurementStdDevs Standard deviations of measurements.
+   * @param dt                 Nominal discretization timestep.
+   */
+  KalmanFilter(LinearSystem<States, Inputs, Outputs>& plant,
+               const std::array<double, States>& stateStdDevs,
+               const std::array<double, Outputs>& measurementStdDevs,
+               units::second_t dt) {
+    m_plant = &plant;
+
+    m_contQ = MakeCovMatrix(stateStdDevs);
+    m_contR = MakeCovMatrix(measurementStdDevs);
+
+    Eigen::Matrix<double, States, States> discA;
+    Eigen::Matrix<double, States, States> discQ;
+    DiscretizeAQTaylor<States>(plant.A(), m_contQ, dt, &discA, &discQ);
+
+    m_discR = DiscretizeR<Outputs>(m_contR, dt);
+
```

Comment:
> Any reason to support non-observable kalman filters?  Personally, I'd just explode and make the user fix it. I'd also recommend documenting the AB vs A.t, C.t trick a bit better.

-- Austin Schuh

I agree with this for the KF. The EKF and UKF will need to stay because certain drivetrain models don't have convergent DAREs (X and Y error covariance are unbounded).